src/gui/src/components/query-builder/ast-tokenize.ts

```diff
@@ -0,0 +1,212 @@
+): AttributeUiNode => {
+  const node: AttributeUiNode = {
+    type: 'attribute',
+    value: postcssNode.value || '',
+    attribute:
+      'attribute' in postcssNode ? postcssNode.attribute || '' : '',
+  }
+
+  if (
+    'operator' in postcssNode &&
+    postcssNode.operator !== undefined
+  ) {
+    node.operator = postcssNode.operator
+  }
+  if (
+    'flag' in postcssNode &&
+    isAttributeFlag((postcssNode as { flag?: unknown }).flag)
+  ) {
+    node.flag = (postcssNode as { flag?: AttributeFlag }).flag
+  }
+  return node
+}
+
+const createContainerNode = (
+  postcssNode: PostcssNode,
+  children: UiNode[],
+): UiNode => {
+  if (postcssNode.type === 'pseudo') {
+    return {
+      type: 'pseudo',
+      value: (postcssNode as { value: string })
+        .value as PseudoUiNode['value'],
+      children,
+    }
+  }
+  return {
+    type: postcssNode.type,
+    value: (postcssNode as { value?: string }).value || '',
+    children,
+  } as UiNode
+}
+
+// Flatten root (selector-list) and insert comma tokens between adjacent selectors
+const flattenRoot = (rootNode: UiNode): UiNode[] | undefined => {
+  if (
+    rootNode.type === 'root' &&
+    'children' in rootNode &&
+    Array.isArray(rootNode.children)
+  ) {
+    const flat: UiNode[] = []
+    const kids = rootNode.children as UiNode[]
+    for (let i = 0; i < kids.length; i++) {
+      const child = kids[i]
+      if (
+        child &&
+        child.type === 'selector' &&
+        'children' in child &&
+        Array.isArray(child.children)
+      ) {
+        flat.push(...(child.children as UiNode[]))
+        const next = kids[i + 1]
+        if (next && next.type === 'selector') {
+          flat.push({ type: 'comma', value: ',' })
+        }
+      } else if (
+        child &&
+        child.type === 'selector' &&
+        (!child.value || child.value === '')
+      ) {
+        continue
+      } else if (child) {
+        flat.push(child)
+      }
+    }
+    return flat.length > 0 ? flat : undefined
+  }
+
+  if (
+    rootNode.type === 'root' &&
+    (!rootNode.value || rootNode.value === '')
+  ) {
+    return undefined
+  }
+  return [rootNode]
+}
+
+// Preserve trailing draft tokens while typing (comma and descendant space)
+const appendDraftSuffixTokens = (
+  query: string,
+  tokens: UiNode[] | undefined,
+): UiNode[] | undefined => {
+  if (!tokens) return tokens
+  let result = tokens
+
+  if (
+    /,\s*$/.test(query) &&
+    result[result.length - 1]?.type !== 'comma'
+  ) {
+    result = [...result, { type: 'comma', value: ',' }]
+  }
+
+  const trimmed = query.replace(/\s+$/, '')
+  const hasTrailingSpace = trimmed.length < query.length
+  if (
+    hasTrailingSpace &&
+    !/[>,~]$/.test(trimmed) &&
+    (result[result.length - 1]?.type !== 'combinator' ||
+      (result[result.length - 1] as UiNode & { value?: string })
+        .value !== ' ')
+  ) {
+    result = [...result, { type: 'combinator', value: ' ' }]
+  }
+
+  return result
+}
```

Comment:
Let's ideally remove the regex checks here, also some checks seem to be duplicated, ideally we're relying only on the ast objects, we don't need to be validating and / or checking the query string at the same time